.github/workflows/ci.yml

```diff
@@ -0,0 +1,166 @@
+          echo "Generated filters: $FILTERS"
+          
+          echo "modules_json=$MODULES_JSON" >> $GITHUB_OUTPUT 
+          echo "projects_json=$PROJECTS_JSON" >> $GITHUB_OUTPUT
+          
+          {
+            echo 'filters<<EOF'
+            echo "$FILTERS"
+            echo EOF
+          } >> $GITHUB_ENV
+      - uses: dorny/paths-filter@v3
+        id: changes
+        with:
+          filters: ${{ env.filters }}
+      - name: Filter JSONs Based on Changes
+        id: filtered
+        uses: actions/github-script@v7
+        env:
+          MODULES_JSON: ${{ steps.discover.outputs.modules_json }}
+          PROJECTS_JSON: ${{ steps.discover.outputs.projects_json }}
+          CHANGE_JSON: ${{ steps.changes.outputs.changes }}
+        with:
+          script: |
+            const modulesJson = JSON.parse(process.env.MODULES_JSON || '[]');
+            const projectsJson = JSON.parse(process.env.PROJECTS_JSON || '[]');
+            const changeJson = JSON.parse(process.env.CHANGE_JSON || '[]');
+            
+            if (process.env.check_only_changed) {
+              console.log(`Filtering modules based on changes: ${JSON.stringify(changeJson)}`);
+              const filteredModules = modulesJson.filter(module => {
+                return changeJson.some(change => change.includes(module));
+              });
+              const filteredProjects = projectsJson.filter(project => {
+                return changeJson.some(change => change.includes(project));
+              });
+              console.log(`Filtered modules: ${JSON.stringify(filteredModules)}`);
+              core.setOutput("modules_json", JSON.stringify(filteredModules));
+              core.setOutput("projects_json", JSON.stringify(filteredProjects));
+            } else {
+              console.log("check_only_changed is false, no filtering applied.");
+              core.setOutput("modules_json", JSON.stringify(modulesJson));
+              core.setOutput("projects_json", JSON.stringify(projectsJson));
+            }
+
+  golangci_lint:
+    name: "golangci-lint"
+    needs: discover_modules
+    runs-on: ubuntu-latest
+    permissions:
+      # Required: allow read access to the content for analysis.
+      contents: read
+      # Optional: allow write access to checks to allow the action to annotate code in the PR.
+      checks: write
+    strategy:
+      matrix:
+        module: ${{ fromJSON(needs.discover_modules.outputs.modules_json) }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          sparse-checkout: |
+            ${{ matrix.module }}
+            .github/config/golangci.yml
+      - name: Setup Go ${{ matrix.go-version }}
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: '${{ github.workspace }}/${{ matrix.module }}/go.mod'
+      - name: golangci-lint ${{ matrix.modules }}
+        uses: golangci/golangci-lint-action@v6
+        with:
+          version: ${{ env.GOLANGCI_LINT_VERSION }}
```

Comment:
why not using the new go 1.24 feature: `tool`s?
e.g. https://github.com/open-component-model/ocm/blob/main/go.mod#L532

Reply:
https://golangci-lint.run/welcome/install/#install-from-sources

There are various caveats and golangci-lint does not recommend installing from source for this reason.

actually the env variable right now isnt used because it uses the latest version of golangci lint